### PR TITLE
Optimized JSON export

### PR DESCRIPTION
**Pull request by uroybd**
I've implemented a better JSON export format that removes redundant lists and objects and introduces the `entries` key containing all the entries.

It also add `drawer` values from highlight so that user can use this piece of metadata to generate desirable output.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8904)

### CODE REVIEW

**Comment by Frenzie** [comment]
Could you share a before/after JSON? Is it exactly as described in https://github.com/koreader/koreader/issues/8890?

**Comment by uroybd** [comment]
> Could you share a before/after JSON? Is it exactly as described in #8890?

Yes. Except 2 things:

1. I've added the `drawer` values in the entires.
2. I couldn't maintain the order in the top level keys (title, exported, etc.) Entries are ordered properly though.

**Comment by uroybd** [comment]
Also, for single book export, we can consider using the `.sdr` folder of a book and timestamped export file.

Or,

In the clipboard folder like now but in `{filename}-{timestamp}.json` like files.

**Comment by uroybd** [comment]
> Also, for single book export, we can consider using the `.sdr` folder of a book and timestamped export file.
> 
> Or,
> 
> In the clipboard folder like now but in `{filename}-{timestamp}.json` like files.

@NiLuJe
Do you have a thought about it? Multiple set of data just appended in a file (without using comma and `[]` to enclose them) creates an invalid JSON file. We can asume that whoever is exporting JSON will use this data in another pipeline.

**Comment by NiLuJe** [comment]
No idea, I don't use the plug-in, and am thoroughly unfamiliar with the code ;).

**Review by pazos — CHANGES_REQUESTED** [review]
Changes are sensible for a single document but don't work for a list of documents.

It isn't your fault as it is known to be broken in the plugin ATM.

Just wondered how the `exportable` table would look with n items before merging all this. Because the whole chunk of code is intended to remove the array of booknotes, but we might end with an array of documents.

Could you refactor your changes outside `exportBooknotesToJSON` function?

This way:

create a new local function `exportToJson(title, t)`

`t` is an array of documents.

To make it work with a single document just do

```lua
    if file then
        local t = {}
        table.insert(t, booknotes)
        exportToJson(title, t)
    end
```

This way you'll obtain a result that is intended to work with multiple documents. If the result is not what you expected you can change it accordly :)

remember that `booknotes` is an array of notes for a document and `t` is an array of booknotes from the whole history so you'll need to iterate over `t` the same way you did for `booknotes`.

```lua

for _, booknotes in ipairs(t) do
    for __, entry in ipairs(booknotes) do
    end
end
```

File: plugins/exporter.koplugin/clip.lua [hunk 1]
```diff
@@ -256,6 +256,7 @@ function MyClipping:parseHighlight(highlights, bookmarks, book)
             clipping.time = self:getTime(item.datetime or "")
             clipping.text = self:getText(item.text)
             clipping.chapter = item.chapter
+            clipping.drawer = item.drawer
```
[the remark this answers is not in the document]
Comment (uroybd) [reply plugins/exporter.koplugin/clip.lua:259]: Actually, it is used. There are four highlight style, right? It is possible to use them to mark differently. For example, I use strikethrough to mark my disapproval of a statement. This piece of metadata exported to JSON will allow me to process my notes more accurately.

**Comment by uroybd** [comment]
> Changes are sensible for a single document but don't work for a list of documents.
> 
> It isn't your fault as it is known to be broken in the plugin ATM.
> 
> Just wondered how the `exportable` table would look with n items before merging all this. Because the whole chunk of code is intended to remove the array of booknotes, but we might end with an array of documents.
> 
> Could you refactor your changes outside `exportBooknotesToJSON` function?
> 
> This way:
> 
> create a new local function `exportToJson(title, t)`
> 
> `t` is an array of documents.
> 
> To make it work with a single document just do
> 
> ```lua
>     if file then
>         local t = {}
>         table.insert(t, booknotes)
>         exportToJson(title, t)
>     end
> ```
> 
> This way you'll obtain a result that is intended to work with multiple documents. If the result is not what you expected you can change it accordly :)
> 
> remember that `booknotes` is an array of notes for a document and `t` is an array of booknotes from the whole history so you'll need to iterate over `t` the same way you did for `booknotes`.
> 
> ```lua
> for _, booknotes in ipairs(t) do
>     for __, entry in ipairs(booknotes) do
>     end
> end
> ```

Here is how I thought about refractor it for multi-document export:

```lua
function Exporter:prepareBooknotesForJSON(booknotes)
    local exportable = {
            title = booknotes.title,
            author = booknotes.author,
            entries = {},
            exported = booknotes.exported,
            file = booknotes.file
    }
    for _, entry in ipairs(booknotes) do
        table.insert(exportable.entries, entry[1])
    end
    return exportable
end

-- This function should handle both multidocument export and single exports.
-- For Single Exports, it will create a JSON file with a object ({}) as root node.
-- For Multidocument export, it will create a JSON file with an array ([]) as root node.
function Exporter:exportToJSON(documents)
    local file = io.open(self.json_clipping_file, "a")
    if file then
        local exportable = {}
        if table.getn(documents) == 1 then
            -- We will handle single document export here.
            exportable = self:prepareBooknotesForJSON(documents[0])
        else
            for _, document in ipairs(documents) do
                table.insert(exportable, self:prepareBooknotesForJSON(document))
            end
        end
        file:write(json.encode(exportable))
        file:write("\n")
        file:close()
    end
end

function Exporter:exportBooknotesToJSON(title, booknotes)
    local file = io.open(self.json_clipping_file, "a")
    if file then
        local exportable = self:prepareBooknotesForJSON(booknotes)
        file:write(json.encode(exportable))
        file:write("\n")
        file:close()
    end
end
```
I'm not touching `exportBooknotesToJSON` right now since I don't know the data pipeline quite well.
The assumption here is, for multi-document exports I'll get a table (array-like) of `booknote` in `exportToJSON`.

**Comment by uroybd** [comment]
@pazos I think, for exporting multi-document, We should not iterate over clipping first. Different targets can handle data differently. Let us feed the exporters the whole clippings and let them decide how they want to handle data by themselves.

**Comment by pazos** [comment]
> @pazos I think, for exporting multi-document, We should not iterate over clippings first. Different targets can handle data differently. Let us feed the exporters the whole clippings and let them decide how they want to handle data by themselves.

If you check the zip attached in https://github.com/koreader/koreader/issues/8755#issuecomment-1030652265 you'll find exactly that. Each exporter has an `export` function that receives a table of booknotes, no matter if it only contains a single element or a bunch of them.

My suggestion was directed at this specific PR, to be merged before the refactor. Because the intent of this PR is just to remove the array of booknotes and replace it with your own dictionary of them. My point was: check what happens when you try to export more than one document because you will end with an array of documents (each one containing a dictionary of entries)

**Comment by uroybd** [comment]
> > @pazos I think, for exporting multi-document, We should not iterate over clippings first. Different targets can handle data differently. Let us feed the exporters the whole clippings and let them decide how they want to handle data by themselves.
> 
> If you check the zip attached in [#8755 (comment)](https://github.com/koreader/koreader/issues/8755#issuecomment-1030652265) you'll find exactly that. Each exporter has an `export` function that receives a table of booknotes, no matter if it only contains a single element or a bunch of them.
> 
> My suggestion was directed at this specific PR, to be merged before the refactor. Because the intent of this PR is just to remove the array of booknotes and replace it with your own dictionary of them. My point was: check what happens when you try to export more than one document because you will end with an array of documents (each one containing a dictionary of entries)

I've just tested. It looks like this:

```json
{
    "exported": {
        "json": true
    },
    "file": "\/home\/uroybd\/Documents\/EBooks\/(Oxford World's Classics) Friedrich Nietzsche, Graham Parkes - Thus Spoke Zarathustra_ A Book for Everyone and Nobody-Oxford University Press (2005).epub",
    "title": "Thus Spoke Zarathustra: A Book for Everyone and Nobody (Oxford World's Classics)",
    "author": "Friedrich Nietzsche",
    "entries": [
        {
            "drawer": "lighten",
            "time": 1647438032,
            "page": 5,
            "text": "pocket-sized hardbacks of the early years contained introductions by Virginia Woolf, T. S. Eliot, Graham Greene",
            "sort": "highlight",
            "chapter": "Copyright Page"
        }
    ]
}
{
    "exported": {
        "json": true
    },
    "file": "\/home\/uroybd\/Documents\/EBooks\/(Cambridge texts in the history of philosophy) Plato._ Howatson, M. C._ Sheffield, Frisbee Candida Cheyenne - Plato, the Symposium-Cambridge University Press (2008)_epub3.epub",
    "title": "The Symposium",
    "author": "Plato\nM. C. Howatson\nFrisbee C. C. Sheffield\nM. C. Howatson",
    "entries": [
        {
            "drawer": "lighten",
            "time": 1647273524,
            "page": 12,
            "text": "aristocratic circles, on this occasion designed to honour Agathon’s victory. Each guest delivers a speech in praise of eros, ‘passionate love’, or ‘desire’.5 The final",
            "sort": "highlight",
            "chapter": "Introduction"
        },
        {
            "drawer": "underscore",
            "time": 1647273534,
            "page": 13,
            "text": "and rich characterisation. During the time between Agathon’s drinking party and its recollection by Apollodorus, the Athenians had lost some of the confidence",
            "sort": "highlight",
            "chapter": "Introduction"
        },
        {
            "drawer": "underscore",
            "time": 1647273423,
            "page": 72,
            "text": "Persians invade Greece and are defeated at Marathon.",
            "note": "Test Note",
            "sort": "highlight",
            "chapter": "Chronology"
        }
    ]
}
```

File: plugins/exporter.koplugin/clip.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (uroybd) [reply plugins/exporter.koplugin/clip.lua:259]: @pazos Just bringing it to your attention. Do you find my reason justified?

**Comment by uroybd** [comment]
@pazos This export is invalid but as invalid as before, only more usable. Let us keep it this way in this PR. Let it be an intermediate thing and let us work on the refactoring you have started soon. :)

**Review by poire-z — COMMENTED** [review]
(Not familiar with exporter plugin, so no real review from me.
But I agree that item.drawer can/should be exported, as it somehow qualifies the highlight and is some valuable information to export.)

**Comment by uroybd** [comment]
Seems like this is the only PR left for the March milestone. Please let me know if I can be of any assistance. :smile:

**Comment by uroybd** [comment]
@pazos
Do I have to do anything else?

**Comment by pazos** [comment]
@uroybd: Sorry, I don't have time to review until tomorrow or sunday. If somebody else wants to do it I'm fine with that :smile: 

FWIW I don't use the plugin at all

**Comment by Frenzie** [comment]
@uroybd In what way was it invalid before?

**Comment by uroybd** [comment]
> @uroybd In what way was it invalid before?

I'm referring to this comment [here](https://github.com/koreader/koreader/issues/8755#issuecomment-1068024857).

Currently, every new export gets appended to that JSON file. But, a JSON file without a `[]` or `{}` as root is invalid.

**Comment by Frenzie** [comment]
So it's basically identical?

**Comment by uroybd** [comment]
Yes.

**Comment by uroybd** [comment]
I have a strategy in mind to improve the current implementation of JSON export. I will work for it if you think it is worthwhile.

1. Will check if it is JSON export, then deal with it separately, with the whole clippings table sent to the handler function instead of individual book's bookmarks.
2. Then, I will read the export file
  a. If it exists:
     1. I'll try to read the content, and decode it to JSON.
       a. If it fails, that means the previous version of the export, therefore will fix it.
  b. Create new exports, will append them to the exportable json.
  c. Dump the JSON to the file.

The new format I have in mind is such:
```json
{
    "version": "json-2.0.0",
    "exports": [
        {...export entries like before}
    ]
}
```

The version number will be used by software intending to use these exports.

**Comment by Frenzie** [comment]
Discussion probably makes more sense in #8755? Anyway, sounds sensible enough to me (assuming you don't mean literally but with the requisite commas ;-) ). But @pazos has a much better picture of this plugin.

**Comment by uroybd** [comment]
> Discussion probably makes more sense in #8755? Anyway, sounds sensible enough to me (assuming you don't mean literally but with the requisite commas ;-) ). But @pazos has a much better picture of this plugin.

Lol, of course with commas. I'm taking my conversation to the issue you've mentioned.

**Pull request merged**